### ISSUE TITLE
Hide markers that overlap each other

### ISSUE THREAD

**Issue by Tvax** [osmdroid/osmdroid#1404]
## Issue Type

[ ] Question
[ ] Bug
[x] Improvement
[ ] Build system related
[x] Performance
[ ] Documentation

## Description and/or steps/code to reproduce the problem
Add a method to the mapView that hides markers overlapping another one. Refresh every time the zoom value change.
So the map will be less loaded and the app will run smoother. And it would be prettier.

**Comment by monsieurtanuki** [comment]
I understand but I'm not convinced: for which reason should this Marker be on top.
Could be solved if z-index was implemented.
In another issue's comment I developed a multi Marker click:
* it looks as usual on the map, with many Markers at the same place
* but when you click, there's a popup with all the Markers below, not just the top one
* I even think I developed a demo about that, about museums in Berlin

Not sure if it would suit your needs.

**Comment by Tvax** [comment]
The problem is that the user can't see the marker icon if he doesn't zoom in. Opening a popup is good idea but doesn't solve the visibility problem on the mapview itself.

The marker on top could be the one which overlaps the less other markers. 
Or the one that is the more in the middle of the boundingBoox of all the overlapping markers.

**Comment by monsieurtanuki** [comment]
@Tvax As an end-user, I want to know if the area is crowded with POIs.

That being said, how do you decide which icon is more relevant? Is that just luck? I mean, if there's the Louvre museum, but its closest café is pixelly better positioned, would that make sense to display the café instead of the museum?

Maybe the root of all that is: are some POIs more relevant than others? I do think so. Next step would be to implement some kind of z-index layers, then possibly avoid to display layers beneath.

A technical problem is that is not that easy to compute quickly at each `draw` the pixels that have already been drawn on by a POI.

**Comment by Tvax** [comment]
I do think that some POIs are more relevant than others as well.
To filter out the less important ones we should send to a callback the list of POIs overlapping each other. This will allow the developer to return the POI that is the most important one for his needs.

To do so, instead of manually checking which POIs are overlapping or being overlapped thanks to their icons, we could let the developer set a radius around each POI manually. So all POIs in this radius would be sent to the callback explained above.

To avoid any duplicate in the callback, we could temporarily save any POIs that has been sent to the callback. And so the variable storing this POIs would be reset every time the user moves on the mapView or change the zoom level.

**Comment by monsieurtanuki** [comment]
@Tvax I'm afraid this callback mechanism is much too expensive performance-wise - especially during the UI thread.

My best functional suggestion would be to attach a minimum zoom level for `Marker`s or marker folders: below that level, the Marker isn't displayed.
The developer would create "layers" of Markers, to be displayed starting at some levels only, and would have total control over their respective relevancy.
Then, I guess it makes sense to have both Central Park and Rockefeller Center icons overlapping, as long as the map is not polluted with the least coffee joint.

**Comment by MKergall** [comment]
If the developer knowing that some markers are more important than others, he can already handle the display order. Not as simple as a z-index, but still feasible. 

- If there is a priority known individually for each Marker, they can be inserted appropriately in the overlay list: lowest first, highest later. 

- If there are few "kind" of Markers: with let say 2 different priority (either "high" or "low"): display 2 FolderOverlay (low first, high second) and push each Marker in the relevant FolderOverlay.

**Comment by monsieurtanuki** [comment]
@MKergall Indeed. I did try to implement z-index, but eventually found out that we'd be better off with a list of folders, the first one being the one with the lowest priority.
